dpcpp/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,449 @@
+    }
+    if constexpr (n_shared_total == 6) {
+        r_sh = workspace + gmem_offset;
+    } else {
+        r_sh = s_sh + sconf.padded_vec_len;
+    }
+    if constexpr (n_shared_total == 7) {
+        r_hat_sh = workspace + gmem_offset;
+    } else {
+        r_hat_sh = r_sh + sconf.padded_vec_len;
+    }
+    if constexpr (n_shared_total == 8) {
+        x_sh = workspace + gmem_offset;
+    } else {
+        x_sh = r_hat_sh + sconf.padded_vec_len;
+    }
+    if constexpr (n_shared_total == 9) {
+        prec_work_sh = workspace + gmem_offset;
+    } else {
+        prec_work_sh = x_sh + sconf.padded_vec_len;
+    }
+
+    // generate preconditioner
+    prec_shared.generate(batch_id, mat_global_entry, prec_work_sh, item_ct1);
+
+    // initialization
+    // rho_old = 1, omega = 1, alpha = 1
+    // compute b norms
+    // copy x from global to shared memory
+    // r = b - A*x
+    // compute residual norms
+    // r_hat = r
+    // p = 0
+    // v = 0
+    initialize<sg_kernel_all>(num_rows, mat_global_entry, b_global_entry,
+                              x_global_entry, rho_old_sh[0], omega_sh[0],
+                              alpha_sh[0], x_sh, r_sh, r_hat_sh, p_sh, v_sh,
+                              norms_rhs_sh[0], norms_res_sh[0], item_ct1);
+    item_ct1.barrier(sycl::access::fence_space::local_space);
+
+    // stopping criterion object
+    StopType stop(tol, norms_rhs_sh);
+
+    int iter = 0;
+    for (; iter < max_iter; iter++) {
+        if (stop.check_converged(norms_res_sh)) {
+            break;
+        }
+
+        // rho_new =  < r_hat , r > = (r_hat)' * (r)
+        if constexpr (sg_kernel_all) {
+            if (sg_id == 0) {
+                single_rhs_compute_dot_sg(num_rows, r_hat_sh, r_sh,
+                                              rho_new_sh[0], item_ct1);
+            }
+            item_ct1.barrier(sycl::access::fence_space::local_space);
+        } else {
+            single_rhs_compute_dot(num_rows, r_hat_sh, r_sh, rho_new_sh[0],
+                                       item_ct1);
+        }
+
+        // beta = (rho_new / rho_old)*(alpha / omega)
+        // p = r + beta*(p - omega * v)
+        update_p(num_rows, rho_new_sh[0], rho_old_sh[0], alpha_sh[0], omega_sh[0],
+                 r_sh, v_sh, p_sh, item_ct1);
+        item_ct1.barrier(sycl::access::fence_space::local_space);
+
+        // p_hat = precond * p
+        prec_shared.apply(num_rows, p_sh, p_hat_sh, item_ct1);
+        item_ct1.barrier(sycl::access::fence_space::local_space);
+
+        // v = A * p_hat
+        simple_apply_kernel(mat_global_entry, p_hat_sh, v_sh, item_ct1);
+        item_ct1.barrier(sycl::access::fence_space::local_space);
+
+        // alpha = rho_new / < r_hat , v>
+        compute_alpha<sg_kernel_all>(num_rows, rho_new_sh[0], r_hat_sh, v_sh,
+                                     alpha_sh[0], item_ct1);
+        // item_ct1.barrier(sycl::access::fence_space::local_space);
```

Comment:
why is that sync not necessary? Same for the commented out one below.

Reply:
re-enabled now. I think it is a bit difficult to clearly say where a sync is necessary. But for correctness, it is best to have they local_space syncs everywhere.